File: ObjectiveC SDK/DeviceStatusViewController.m
//
//  DeviceStatusViewController.m
//  ObjectiveC SDK
//
//  Created by Yuji on 2016/**/**.
//  Copyright © 2016年 Star Micronics. All rights reserved.
//

#import "DeviceStatusViewController.h"

#import "AppDelegate.h"

#import <StarIO/SMPort.h>

typedef NS_ENUM(NSUInteger, CellParamIndex) {
    CellParamIndexTitleIndex = 0,
    CellParamIndexDetailIndex,
    CellParamIndexColorIndex
};

@interface DeviceStatusViewController ()

@property (nonatomic) NSMutableArray *cellArray;

@property (nonatomic) BOOL didAppear;

- (void)refreshDeviceStatus;

@end

@implementation DeviceStatusViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    [self appendRefreshButton:@selector(refreshDeviceStatus)];
    
    _cellArray = [[NSMutableArray alloc] init];
    
    _didAppear = NO;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    
    if (_didAppear == NO) {
        [self refreshDeviceStatus];
        
        _didAppear = YES;
    }
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    // Return the number of sections.
    
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    // Return the number of rows in the section.
    
    return _cellArray.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *CellIdentifier = @"UITableViewCellStyleValue1";
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleValue1 reuseIdentifier:CellIdentifier];
    }
    
    if (cell != nil) {
        NSArray *cellParam = _cellArray[indexPath.row];
        
        cell.      textLabel.text      = cellParam[CellParamIndexTitleIndex];
        cell.detailTextLabel.text      = cellParam[CellParamIndexDetailIndex];
        cell.detailTextLabel.textColor = cellParam[CellParamIndexColorIndex];
    }
    
    return cell;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    return @"Contents";
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}

- (void)refreshDeviceStatus {
    BOOL result = NO;
    
    self.blind = YES;
    
    [_cellArray removeAllObjects];
    
    SMPort *port = nil;
    
    @try {
        while (YES) {
            port = [SMPort getPort:[AppDelegate getPortName] :[AppDelegate getPortSettings] :10000];     // 10000mS!!!
            
            if (port == nil) {
                break;
            }
            
            StarPrinterStatus_2 printerStatus;
            
            [port getParsedStatus:&printerStatus :2];
            
            if (printerStatus.offline == SM_TRUE) {
                [_cellArray addObject:@[@"Online", @"Offline", [UIColor redColor]]];
            }
            else {
                [_cellArray addObject:@[@"Online", @"Online",  [UIColor blueColor]]];
            }
            
            if (printerStatus.coverOpen == SM_TRUE) {
                [_cellArray addObject:@[@"Cover", @"Open",   [UIColor redColor]]];
            }
            else {
                [_cellArray addObject:@[@"Cover", @"Closed", [UIColor blueColor]]];
            }
            
            if (printerStatus.receiptPaperEmpty == SM_TRUE) {
                [_cellArray addObject:@[@"Paper", @"Empty", [UIColor redColor]]];
            }
            else if (printerStatus.receiptPaperNearEmptyInner == SM_TRUE ||
                     printerStatus.receiptPaperNearEmptyOuter == SM_TRUE) {
                [_cellArray addObject:@[@"Paper", @"Near Empty", [UIColor orangeColor]]];
            }
            else {
                [_cellArray addObject:@[@"Paper", @"Ready",      [UIColor blueColor]]];
            }
            
            if ([AppDelegate getCashDrawerOpenActiveHigh] == YES) {
                if (printerStatus.compulsionSwitch == SM_TRUE) {
                    [_cellArray addObject:@[@"Cash Drawer", @"Open",   [UIColor redColor]]];
                }
                else {
                    [_cellArray addObject:@[@"Cash Drawer", @"Closed", [UIColor blueColor]]];
                }
            }
            else {
                if (printerStatus.compulsionSwitch == SM_TRUE) {
                    [_cellArray addObject:@[@"Cash Drawer", @"Closed", [UIColor blueColor]]];
                }
                else {
                    [_cellArray addObject:@[@"Cash Drawer", @"Open",   [UIColor redColor]]];
                }
            }
            
            if (printerStatus.overTemp == SM_TRUE) {
                [_cellArray addObject:@[@"Head Temperature", @"High",   [UIColor redColor]]];
            }
            else {
                [_cellArray addObject:@[@"Head Temperature", @"Normal", [UIColor blueColor]]];
            }
            
            if (printerStatus.unrecoverableError == SM_TRUE) {
                [_cellArray addObject:@[@"Non Recoverable Error", @"Occurs", [UIColor redColor]]];
            }
            else {
                [_cellArray addObject:@[@"Non Recoverable Error", @"Ready",  [UIColor blueColor]]];
            }
            
            if (printerStatus.cutterError == SM_TRUE) {
                [_cellArray addObject:@[@"Paper Cutter", @"Error", [UIColor redColor]]];
            }
            else {
                [_cellArray addObject:@[@"Paper Cutter", @"Ready", [UIColor blueColor]]];
            }
            
            if (printerStatus.headThermistorError == SM_TRUE) {
                [_cellArray addObject:@[@"Head Thermistor", @"Error",  [UIColor redColor]]];
            }
            else {
                [_cellArray addObject:@[@"Head Thermistor", @"Normal", [UIColor blueColor]]];
            }
            
            if (printerStatus.voltageError == SM_TRUE) {
                [_cellArray addObject:@[@"Voltage", @"Error",  [UIColor redColor]]];
            }
            else {
                [_cellArray addObject:@[@"Voltage", @"Normal", [UIColor blueColor]]];
            }
            
            if (printerStatus.etbAvailable == SM_TRUE) {
                [_cellArray addObject:@[@"ETB Counter", [NSString stringWithFormat:@"%d", printerStatus.etbCounter], [UIColor blueColor]]];
            }
            
            result = YES;
            break;
        }
    }
    @catch (PortException *exception) {
    }
    @finally {
        if (port != nil) {
            [SMPort releasePort:port];
        }
    }
    
    if (result == NO) {
        UIAlertView *alertView = [[UIAlertView alloc] initWithTitle:@"Fail to Open Port" message:nil delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil];
        
        [alertView show];
    }
    
    [self.tableView reloadData];
    
    self.blind = NO;
}

@end

File: Software/StarPRNT SDK/Swift SDK/Swift SDK/DeviceStatusViewController.swift
//
//  DeviceStatusViewController.swift
//  Swift SDK
//
//  Created by Yuji on 2016/**/**.
//  Copyright © 2016年 Star Micronics. All rights reserved.
//

import UIKit

class DeviceStatusViewController: CommonViewController, UITableViewDelegate, UITableViewDataSource, UIAlertViewDelegate {
    enum CellParamIndex: Int {
        case TitleIndex = 0
        case DetailIndex
        case ColorIndex
    }
    
    @IBOutlet weak var tableView: UITableView!
    
    var cellArray: NSMutableArray!
    
    var didAppear: Bool!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
//      self.appendRefreshButton                                    ("refreshDeviceStatus")
        self.appendRefreshButton(#selector(DeviceStatusViewController.refreshDeviceStatus))
        
        self.cellArray = NSMutableArray()
        
        self.didAppear = false
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    override func viewDidAppear(animated: Bool) {
        super.viewDidAppear(animated)
        
        if didAppear == false {
            self.refreshDeviceStatus()
            
            self.didAppear = true
        }
    }
    
    func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.cellArray.count
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cellIdentifier: String = "UITableViewCellStyleValue1"
        
        var cell: UITableViewCell! = self.tableView.dequeueReusableCellWithIdentifier(cellIdentifier)
        
        if cell == nil {
            cell = UITableViewCell(style: UITableViewCellStyle.Value1, reuseIdentifier: cellIdentifier)
        }
        
        if cell != nil {
            let cellParam: [AnyObject] = self.cellArray[indexPath.row] as! [AnyObject]
            
            cell      .textLabel!.text      = cellParam[CellParamIndex.TitleIndex .rawValue] as? String
            cell.detailTextLabel!.text      = cellParam[CellParamIndex.DetailIndex.rawValue] as? String
            cell.detailTextLabel!.textColor = cellParam[CellParamIndex.ColorIndex .rawValue] as! UIColor
        }
        
        return cell
    }
    
    func tableView(tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        return "Contents"
    }
    
    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        self.tableView.deselectRowAtIndexPath(indexPath, animated: true)
    }
    
    func refreshDeviceStatus() {
        var result: Bool = false
        
        self.blind = true
        
        defer {
            self.blind = false
        }
        
        self.cellArray.removeAllObjects()
        
        var error: NSError?
        
        while true {
            guard let port: SMPort = SMPort.getPort(AppDelegate.getPortName(), AppDelegate.getPortSettings(), 10000) else {     // 10000mS!!!
                break
            }
            
            defer {
                SMPort.releasePort(port)
            }
            
            var printerStatus: StarPrinterStatus_2 = StarPrinterStatus_2()
            
            port.getParsedStatus(&printerStatus, 2, &error)
            
            if error != nil {
                break
            }
            
            if printerStatus.offline == sm_true {
                self.cellArray.addObject(["Online", "Offline", UIColor.redColor()])
            }
            else {
                self.cellArray.addObject(["Online",  "Online",  UIColor.blueColor()])
            }
            
            if printerStatus.coverOpen == sm_true {
                self.cellArray.addObject(["Cover", "Open",   UIColor.redColor()])
            }
            else {
                self.cellArray.addObject(["Cover", "Closed", UIColor.blueColor()])
            }
            
            if printerStatus.receiptPaperEmpty == sm_true {
                self.cellArray.addObject(["Paper", "Empty", UIColor.redColor()])
            }
            else if printerStatus.receiptPaperNearEmptyInner == sm_true ||
                    printerStatus.receiptPaperNearEmptyOuter == sm_true {
                self.cellArray.addObject(["Paper", "Near Empty", UIColor.orangeColor()])
            }
            else {
                self.cellArray.addObject(["Paper", "Ready",      UIColor.blueColor()])
            }
            
            if AppDelegate.getCashDrawerOpenActiveHigh() == true {
                if printerStatus.compulsionSwitch == sm_true {
                    self.cellArray.addObject(["Cash Drawer", "Open",   UIColor.redColor()])
                }
                else {
                    self.cellArray.addObject(["Cash Drawer", "Closed", UIColor.blueColor()])
                }
            }
            else {
                if printerStatus.compulsionSwitch == sm_true {
                    self.cellArray.addObject(["Cash Drawer", "Closed", UIColor.blueColor()])
                }
                else {
                    self.cellArray.addObject(["Cash Drawer", "Open",   UIColor.redColor()])
                }
            }
            
            if printerStatus.overTemp == sm_true {
                self.cellArray.addObject(["Head Temperature", "High",   UIColor.redColor()])
            }
            else {
                self.cellArray.addObject(["Head Temperature", "Normal", UIColor.blueColor()])
            }
            
            if printerStatus.unrecoverableError == sm_true {
                self.cellArray.addObject(["Non Recoverable Error", "Occurs", UIColor.redColor()])
            }
            else {
                self.cellArray.addObject(["Non Recoverable Error", "Ready",  UIColor.blueColor()])
            }
            
            if printerStatus.cutterError == sm_true {
                self.cellArray.addObject(["Paper Cutter", "Error", UIColor.redColor()])
            }
            else {
                self.cellArray.addObject(["Paper Cutter", "Ready", UIColor.blueColor()])
            }
            
            if printerStatus.headThermistorError == sm_true {
                self.cellArray.addObject(["Head Thermistor", "Error",  UIColor.redColor()])
            }
            else {
                self.cellArray.addObject(["Head Thermistor", "Normal", UIColor.blueColor()])
            }
            
            if printerStatus.voltageError == sm_true {
                self.cellArray.addObject(["Voltage", "Error",  UIColor.redColor()])
            }
            else {
                self.cellArray.addObject(["Voltage", "Normal", UIColor.blueColor()])
            }
            
            if printerStatus.etbAvailable == sm_true {
                self.cellArray.addObject(["ETB Counter", String(format: "%d", printerStatus.etbCounter), UIColor.blueColor()])
            }
            
            result = true
            break
        }
        
        if result == false {
            let alertView: UIAlertView = UIAlertView.init(title: "Fail to Open Port", message: "", delegate: nil, cancelButtonTitle: "OK")
            
            alertView.show()
        }
        
        self.tableView.reloadData()
    }
}
